# DPhi Pandas Assignment Details

Dear Learner, 

Congratulations, you have studied Pandas all the way through the end of this bootcamp!!! This is the final graded hands-on exercise where you will be accomplishing the below mentioned task.

# Step 1: Select a Dataset of your choice

*   https://www.kdnuggets.com/datasets/index.html
*   https://dphi.tech/challenges/#past 
* https://pub.towardsai.net/best-datasets-for-machine-learning-data-science-computer-vision-nlp-ai-c9541058cf4f  
* https://medium.com/analytics-vidhya/top-100-open-source-datasets-for-data-science-cd5a8d67cc3d


or you can pick any dataset of your interest on the internet and add reference to it

## Selected Dataset
https://github.com/dphi-official/Datasets/blob/master/avocado.csv

# Evaluation Criteria

Your submission will be evaluated using the following criteria:

* You must ensure that the dataset is tabular.  
* You must ask and answer at least 3 questions about the dataset
*   Your submission must include explanations using markdown cells, apart from the code.
*   Your work must not be plagiarized i.e. copied for somewhere else.

# Step 2: Perform Data Cleaning/Preparation & Analysis

1. Create data frame(s) from CSV files
2. Perform index operations on data frame(s)
3. Apply Group, merge and aggregate data frame(s)
4. Check for missing and invalid values in data
5. Propose a way to treat the missing and invalid values and implement the same on the dataframe
5. Compute the mean, sum, range and other interesting statistics for numeric columns


1. **Create data frame(s) from CSV files:**
**Only import pandas read csv file and setting table to see better than default tables in Jupyter notebook** <br>

In [1]:
import pandas as pd
df_avocado = pd.read_csv('data/avocado.csv')

In [2]:
from IPython.core.display import HTML
display(HTML("<style>pre { white-space: pre !important; }</style>"))

2. **Perform index operations on data frame(s)**

**Use method head to see all columns names and some values**

In [6]:
df_avocado.head()

Unnamed: 0        Date  AveragePrice  Total Volume     4046       4225  \
0           0  2015-12-27          1.33      64236.62  1036.74   54454.85   
1           1  2015-12-20          1.35      54876.98   674.28   44638.81   
2           2  2015-12-13          0.93     118220.22   794.70  109149.67   
3           3  2015-12-06          1.08      78992.15  1132.00   71976.41   
4           4  2015-11-29          1.28      51039.60   941.48   43838.39   

     4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  \
0   48.16     8696.87     8603.62       93.25          0.0  conventional   
1   58.33     9505.56     9408.07       97.49          0.0  conventional   
2  130.50     8145.35     8042.21      103.14          0.0  conventional   
3   72.58     5811.16     5677.40      133.76          0.0  conventional   
4   75.78     6183.95     5986.26      197.69          0.0  conventional   

   year  region  
0  2015  Albany  
1  2015  Albany  
2  2015  Albany  
3  2015  Albany  
4  2015  Albany

**Use of iloc to see other region as BaltimoreWashington, Atlanta and not only see the region Albany**

In [11]:
print(df_avocado.iloc[70:75, 13])
print(df_avocado.iloc[150:155, 13])

70    Atlanta
71    Atlanta
72    Atlanta
73    Atlanta
74    Atlanta
Name: region, dtype: object
150    BaltimoreWashington
151    BaltimoreWashington
152    BaltimoreWashington
153    BaltimoreWashington
154    BaltimoreWashington
Name: region, dtype: object


**Use method tail to see amount of records exist in the csv**

In [13]:
df_avocado.tail()

Unnamed: 0        Date  AveragePrice  Total Volume     4046     4225  \
18244           7  2018-02-04          1.63      17074.83  2046.96  1529.20   
18245           8  2018-01-28          1.71      13888.04  1191.70  3431.50   
18246           9  2018-01-21          1.87      13766.76  1191.92  2452.79   
18247          10  2018-01-14          1.93      16205.22  1527.63  2981.04   
18248          11  2018-01-07          1.62      17489.58  2894.77  2356.13   

         4770  Total Bags  Small Bags  Large Bags  XLarge Bags     type  year  \
18244    0.00    13498.67    13066.82      431.85          0.0  organic  2018   
18245    0.00     9264.84     8940.04      324.80          0.0  organic  2018   
18246  727.94     9394.11     9351.80       42.31          0.0  organic  2018   
18247  727.01    10969.54    10919.54       50.00          0.0  organic  2018   
18248  224.53    12014.15    11988.14       26.01          0.0  organic  2018   

                 region  
18244  WestTexNewMexico  
18245  WestTexNewMexico  
18246  WestTexNewMexico  
18247  WestTexNewMexico  
18248  WestTexNewMexico

**Use loc to see other values different that 0 in the column XLarge Bags, Unfortunately it doesn't same result with column named type where the value still "conventional"**

In [18]:
df_avocado.loc[50:60,['XLarge Bags','type']]

XLarge Bags          type
50         0.00  conventional
51         0.00  conventional
52         0.00  conventional
53         0.00  conventional
54        13.53  conventional
55         2.71  conventional
56        24.41  conventional
57         0.00  conventional
58         0.00  conventional
59        24.60  conventional
60         8.22  conventional

3. **Apply Group, merge and aggregate data frame(s)**

**Add column with rounded price of 1 decimal to easier the calculates**

In [54]:
df_general_price = df_avocado.AveragePrice
df_general_price = round(df_general_price,1)

df_avocado['generalAveragePrice']= df_general_price.values
df_avocado

Unnamed: 0        Date  AveragePrice  Total Volume     4046       4225  \
0               0  2015-12-27          1.33      64236.62  1036.74   54454.85   
1               1  2015-12-20          1.35      54876.98   674.28   44638.81   
2               2  2015-12-13          0.93     118220.22   794.70  109149.67   
3               3  2015-12-06          1.08      78992.15  1132.00   71976.41   
4               4  2015-11-29          1.28      51039.60   941.48   43838.39   
...           ...         ...           ...           ...      ...        ...   
18244           7  2018-02-04          1.63      17074.83  2046.96    1529.20   
18245           8  2018-01-28          1.71      13888.04  1191.70    3431.50   
18246           9  2018-01-21          1.87      13766.76  1191.92    2452.79   
18247          10  2018-01-14          1.93      16205.22  1527.63    2981.04   
18248          11  2018-01-07          1.62      17489.58  2894.77    2356.13   

         4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  \
0       48.16     8696.87     8603.62       93.25          0.0  conventional   
1       58.33     9505.56     9408.07       97.49          0.0  conventional   
2      130.50     8145.35     8042.21      103.14          0.0  conventional   
3       72.58     5811.16     5677.40      133.76          0.0  conventional   
4       75.78     6183.95     5986.26      197.69          0.0  conventional   
...       ...         ...         ...         ...          ...           ...   
18244    0.00    13498.67    13066.82      431.85          0.0       organic   
18245    0.00     9264.84     8940.04      324.80          0.0       organic   
18246  727.94     9394.11     9351.80       42.31          0.0       organic   
18247  727.01    10969.54    10919.54       50.00          0.0       organic   
18248  224.53    12014.15    11988.14       26.01          0.0       organic   

       year            region  generalAveragePrice  
0      2015            Albany                  1.3  
1      2015            Albany                  1.4  
2      2015            Albany                  0.9  
3      2015            Albany                  1.1  
4      2015            Albany                  1.3  
...     ...               ...                  ...  
18244  2018  WestTexNewMexico                  1.6  
18245  2018  WestTexNewMexico                  1.7  
18246  2018  WestTexNewMexico                  1.9  
18247  2018  WestTexNewMexico                  1.9  
18248  2018  WestTexNewMexico                  1.6  

[18249 rows x 15 columns]

4. **Check for missing and invalid values in data**

In [47]:
df_avocado.isnull().sum()

Unnamed: 0      0
Date            0
AveragePrice    0
Total Volume    0
4046            0
4225            0
4770            0
Total Bags      0
Small Bags      0
Large Bags      0
XLarge Bags     0
type            0
year            0
region          0
dtype: int64

5. **Propose a way to treat the missing and invalid values and implement the same on the dataframe**
**In this case there aren't values NaN in any columns then I don't treat**

6. **Compute the mean, sum, range and other interesting statistics for numeric columns**

**First apllication method info cause I want exclude columns of types numeric to apply aggregate function in columns numerics**

In [34]:
df_avocado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18249 entries, 0 to 18248
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    18249 non-null  int64  
 1   Date          18249 non-null  object 
 2   AveragePrice  18249 non-null  float64
 3   Total Volume  18249 non-null  float64
 4   4046          18249 non-null  float64
 5   4225          18249 non-null  float64
 6   4770          18249 non-null  float64
 7   Total Bags    18249 non-null  float64
 8   Small Bags    18249 non-null  float64
 9   Large Bags    18249 non-null  float64
 10  XLarge Bags   18249 non-null  float64
 11  type          18249 non-null  object 
 12  year          18249 non-null  int64  
 13  region        18249 non-null  object 
dtypes: float64(9), int64(2), object(3)
memory usage: 1.9+ MB


In [37]:
df_avocado_numerics = df_avocado.select_dtypes(include=['int64','float64'])
df_avocado_numerics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18249 entries, 0 to 18248
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    18249 non-null  int64  
 1   AveragePrice  18249 non-null  float64
 2   Total Volume  18249 non-null  float64
 3   4046          18249 non-null  float64
 4   4225          18249 non-null  float64
 5   4770          18249 non-null  float64
 6   Total Bags    18249 non-null  float64
 7   Small Bags    18249 non-null  float64
 8   Large Bags    18249 non-null  float64
 9   XLarge Bags   18249 non-null  float64
 10  year          18249 non-null  int64  
dtypes: float64(9), int64(2)
memory usage: 1.5 MB


**Apply mean in average price and Total volume** 

In [39]:
df_avocado_numerics.AveragePrice.mean()

1.405978409775878

In [40]:
df_avocado_numerics['Total Volume'].mean()

850644.0130089321

**Apply unique in region to see all regions also with colum type**

In [41]:
df_avocado.region.unique()

array(['Albany', 'Atlanta', 'BaltimoreWashington', 'Boise', 'Boston',
       'BuffaloRochester', 'California', 'Charlotte', 'Chicago',
       'CincinnatiDayton', 'Columbus', 'DallasFtWorth', 'Denver',
       'Detroit', 'GrandRapids', 'GreatLakes', 'HarrisburgScranton',
       'HartfordSpringfield', 'Houston', 'Indianapolis', 'Jacksonville',
       'LasVegas', 'LosAngeles', 'Louisville', 'MiamiFtLauderdale',
       'Midsouth', 'Nashville', 'NewOrleansMobile', 'NewYork',
       'Northeast', 'NorthernNewEngland', 'Orlando', 'Philadelphia',
       'PhoenixTucson', 'Pittsburgh', 'Plains', 'Portland',
       'RaleighGreensboro', 'RichmondNorfolk', 'Roanoke', 'Sacramento',
       'SanDiego', 'SanFrancisco', 'Seattle', 'SouthCarolina',
       'SouthCentral', 'Southeast', 'Spokane', 'StLouis', 'Syracuse',
       'Tampa', 'TotalUS', 'West', 'WestTexNewMexico'], dtype=object)

In [43]:
df_avocado.type.unique()

array(['conventional', 'organic'], dtype=object)

**Now use functions value_counts to see amount of use in columns not numerics**

In [44]:
df_avocado.region.value_counts()

Albany                 338
Sacramento             338
Northeast              338
NorthernNewEngland     338
Orlando                338
Philadelphia           338
PhoenixTucson          338
Pittsburgh             338
Plains                 338
Portland               338
RaleighGreensboro      338
RichmondNorfolk        338
Roanoke                338
SanDiego               338
Atlanta                338
SanFrancisco           338
Seattle                338
SouthCarolina          338
SouthCentral           338
Southeast              338
Spokane                338
StLouis                338
Syracuse               338
Tampa                  338
TotalUS                338
West                   338
NewYork                338
NewOrleansMobile       338
Nashville              338
Midsouth               338
BaltimoreWashington    338
Boise                  338
Boston                 338
BuffaloRochester       338
California             338
Charlotte              338
Chicago                338
C

In [45]:
df_avocado.type.value_counts()

conventional    9126
organic         9123
Name: type, dtype: int64

# Step 3: Exploration & Inference

1. Ask at least 3 interesting questions about your dataset
2. Answer the questions by computing the results using Pandas library
3. Add concluding remarks


## Questions
1. Which region has the highest average price? and which region has the lowest average price?
2. Which region has the most and least number of small, large, and extra-large backpacks?
3. which has the highest average price between 2 groups of types of avocados conventional and organic? that is, who has the best business.

## Question 1
**First find the max and min average price to use as filter** <br>
**next, my condition find records that it has min value or max value** 

## Solution with Pandas

In [43]:
maxvalue_AveragePrice = df_avocado.AveragePrice.max()
minvalue_AveragePrice = df_avocado.AveragePrice.min()

print("Max:",maxvalue_AveragePrice)
print("Min:",minvalue_AveragePrice)

Max: 3.25
Min: 0.44


In [12]:
condition = (df_avocado.AveragePrice == maxvalue_AveragePrice) | (df_avocado.AveragePrice == minvalue_AveragePrice)
df_avocado[condition][['AveragePrice','region']]

AveragePrice            region
14125          3.25      SanFrancisco
15261          0.44  CincinnatiDayton

**concluding remarks:The answer are San Francisco with the max value of Average Price with 3.25 and min value in CincinnatiDayton with 0.44**

## Question 2
**First find the max and min amount for small,large and Xlarge backpacks to use as filter** <br>
**next, my condition find in base each amount find** 

In [18]:
# Define function to get max and min values with column specified 
def get_amount_max_and_min(columnName:str):
    return df_avocado[columnName].max(), df_avocado[columnName].min()
    

In [25]:
# Small Bags
maxAmount_small , minAmount_small = get_amount_max_and_min("Small Bags")
# Large Bags
maxAmount_large , minAmount_large = get_amount_max_and_min("Large Bags")
# XLarge Bags
maxAmount_xLarge , minAmount_xLarge = get_amount_max_and_min("XLarge Bags")

print("Max Small:",maxAmount_small)
print("Min Small:",minAmount_small,end="\n\n")

print("Max Large:",maxAmount_large)
print("Min Large:",minAmount_large,end="\n\n")

print("Max XLarge:",maxAmount_xLarge)
print("Min XLarge:",minAmount_xLarge)

Max Small: 13384586.8
Min Small: 0.0

Max Large: 5719096.61
Min Large: 0.0

Max XLarge: 551693.65
Min XLarge: 0.0


**One function to select only records that regions with max and min amount of bag depending column name**

In [27]:
def get_record_max_min_amount_bag(columnName:str,values: tuple):
    condition = (df_avocado[columnName] == values[0]) | (df_avocado[columnName] == values[1])
    return df_avocado[condition][[columnName,'region']]

**For each type of bags (Small,Large and XLarge) get records and get region name with function unique to see best**

In [39]:
df_small = get_record_max_min_amount_bag("Small Bags",(maxAmount_small , minAmount_small))
regionsNamesMaxSmall = df_small[df_small["Small Bags"]==maxAmount_small].region.unique()
print("Max in Small:",regionsNamesMaxSmall)

regionsNamesMinSmall = df_sma**One function to select only records that regions with max and min amount of bag depending column name**ll[df_small["Small Bags"]==minAmount_small].region.unique()
print("Min in Small:",regionsNamesMinSmall)

Max in Small: ['TotalUS']
Min in Small: ['Atlanta' 'Boise' 'Columbus' 'LasVegas' 'Louisville' 'Portland'
 'SanFrancisco' 'Seattle' 'Spokane' 'StLouis' 'Denver' 'Jacksonville']


In [41]:
df_large = get_record_max_min_amount_bag("Large Bags",(maxAmount_large , minAmount_large))
regionsNamesMaxLarge = df_large[df_large["Large Bags"] == maxAmount_large].region.unique()
print("Max in Large:",regionsNamesMaxLarge,end="\n\n")

regionsNamesMinLarge = df_large[df_large["Large Bags"] == minAmount_large].region.unique()
print("Min in Large:",regionsNamesMinLarge)

Max in Large: ['TotalUS']

Min in Large: ['Boise' 'BuffaloRochester' 'NorthernNewEngland' 'PhoenixTucson'
 'Portland' 'RichmondNorfolk' 'Roanoke' 'Sacramento' 'Seattle' 'Spokane'
 'StLouis' 'Syracuse' 'Boston' 'HartfordSpringfield' 'Albany' 'Atlanta'
 'BaltimoreWashington' 'California' 'Charlotte' 'Chicago' 'DallasFtWorth'
 'GrandRapids' 'HarrisburgScranton' 'Houston' 'Jacksonville' 'LasVegas'
 'LosAngeles' 'MiamiFtLauderdale' 'NewOrleansMobile' 'NewYork' 'Orlando'
 'Philadelphia' 'Pittsburgh' 'RaleighGreensboro' 'SanDiego' 'SanFrancisco'
 'SouthCarolina' 'Tampa' 'WestTexNewMexico' 'Louisville' 'Columbus']


In [42]:
df_xLarge = get_record_max_min_amount_bag("XLarge Bags",(maxAmount_xLarge , minAmount_xLarge))
regionsNamesMaxxLarge = df_xLarge[df_xLarge["XLarge Bags"]==maxAmount_xLarge].region.unique()
print("Max in XLarge:",regionsNamesMaxxLarge,end="\n\n")

regionsNamesMinxLarge = df_xLarge[df_xLarge["XLarge Bags"]==minAmount_xLarge].region.unique()
print("Min in XLarge:",regionsNamesMinxLarge)

Max in XLarge: ['TotalUS']

Min in XLarge: ['Albany' 'Atlanta' 'BaltimoreWashington' 'Boise' 'Boston'
 'BuffaloRochester' 'Charlotte' 'Chicago' 'DallasFtWorth' 'Denver'
 'HarrisburgScranton' 'HartfordSpringfield' 'Houston' 'Indianapolis'
 'Jacksonville' 'LasVegas' 'Louisville' 'MiamiFtLauderdale' 'Nashville'
 'NewOrleansMobile' 'NewYork' 'Northeast' 'NorthernNewEngland' 'Orlando'
 'Philadelphia' 'PhoenixTucson' 'Pittsburgh' 'Portland'
 'RaleighGreensboro' 'RichmondNorfolk' 'Roanoke' 'Sacramento' 'SanDiego'
 'SanFrancisco' 'Seattle' 'SouthCarolina' 'SouthCentral' 'Southeast'
 'Spokane' 'StLouis' 'Syracuse' 'Tampa' 'West' 'WestTexNewMexico'
 'CincinnatiDayton' 'Columbus' 'Detroit' 'Midsouth' 'Plains' 'California'
 'GrandRapids' 'GreatLakes' 'LosAngeles' 'TotalUS']


**concluding remarks:The answers are before cells where note that the max amount only one region has** <br>
**On the other hand, with the minimum amount, there is more than one region.**

## Question 3
**Only select in dataframe the type of avocado (conventional, organic) and use function max to find average price**

In [49]:
maxPriceConventional = df_avocado[df_avocado.type == "conventional"].AveragePrice.max()
maxPriceConventional

2.22

In [50]:
maxPriceOrganic = df_avocado[df_avocado.type == "organic"].AveragePrice.max()
maxPriceOrganic

3.25

In [51]:
maxPriceOrganic  > maxPriceConventional

True

**concluding remarks:The type organic has higher price than price of conventional**